crates/core/src/factories/core_fetch/test.rs

```diff
@@ -0,0 +1,600 @@
+        agent: Some(agent_1.clone()),
+        url: Some(Some(Url::from_str("wss://127.0.0.1:1").unwrap())),
+        space: Some(space_id.clone()),
+        ..Default::default()
+    }
+    .build();
+    let agent_info_2 = AgentBuild {
+        agent: Some(agent_2.clone()),
+        url: Some(Some(Url::from_str("wss://127.0.0.1:2").unwrap())),
+        space: Some(space_id.clone()),
+        ..Default::default()
+    }
+    .build();
+    let agent_info_3 = AgentBuild {
+        agent: Some(agent_3.clone()),
+        url: Some(Some(Url::from_str("wss://127.0.0.1:3").unwrap())),
+        space: Some(space_id.clone()),
+        ..Default::default()
+    }
+    .build();
+    let agent_url_1 = agent_info_1.url.clone().unwrap();
+    let agent_url_2 = agent_info_2.url.clone().unwrap();
+    let agent_url_3 = agent_info_3.url.clone().unwrap();
+    peer_store
+        .insert(vec![agent_info_1, agent_info_2, agent_info_3])
+        .await
+        .unwrap();
+
+    let fetch = CoreFetch::new(
+        config.clone(),
+        space_id.clone(),
+        peer_store.clone(),
+        mock_transport.clone(),
+    );
+
+    // Add all ops to the queue.
+    futures::future::join_all([
+        fetch.add_ops(op_list_1.clone(), agent_1.clone()),
+        fetch.add_ops(op_list_2.clone(), agent_2.clone()),
+        fetch.add_ops(op_list_3.clone(), agent_3.clone()),
+    ])
+    .await;
+
+    // Wait for one request for each agent.
+    let expected_agent_url = [agent_url_1, agent_url_2, agent_url_3];
+    let expected_agents = [agent_1, agent_2, agent_3];
+    tokio::time::timeout(Duration::from_millis(100), async {
+        loop {
+            tokio::time::sleep(Duration::from_millis(1)).await;
+            let requests_sent =
+                mock_transport.requests_sent.lock().unwrap().clone();
+            let request_destinations = requests_sent
+                .iter()
+                .map(|(_, agent_id)| agent_id)
+                .collect::<Vec<_>>();
+            if expected_agent_url
+                .iter()
+                .all(|agent| request_destinations.contains(&agent))
+            {
+                // Check all agents are on cool-down_list.
+                let cool_down_list =
+                    &mut fetch.state.lock().unwrap().cool_down_list;
+                if expected_agents
+                    .iter()
+                    .all(|agent| cool_down_list.is_agent_cooling_down(agent))
+                {
+                    break;
+                }
+            }
+        }
+    })
+    .await
+    .unwrap();
+
+    // Give time to remove op id from set.
+    tokio::time::sleep(Duration::from_millis(1)).await;
+
+    // Op should have been removed from ops to fetch.
+    assert!(
+        fetch.state.lock().unwrap().ops.is_empty(),
+        "ops are {:?}",
+        fetch.state.lock().unwrap().ops.len()
+    );
+}
+
+#[tokio::test(flavor = "multi_thread")]
+async fn agent_cooling_down_is_removed_from_list() {
```

Comment:
Are they actually removed from anywhere? Or just tracked as being on cool down? Just checking that the name of the test is accurate, the test content looks good but I'm wondering if I'm missing something here

Reply:
They are removed, but no explicitly. As it is right now, `is_agent_cooling_down` removes the agent from the list if the cool-down interval has elapsed. But see below for an extended discussion.